inconsistency in nonterminal/rule.hpp BOOST_SPIRIT_{DEFINE,DECLARE,INSTANTIATE}

OOPS.  See next comment.

> Also it seems that after #347 the debug will be called twice.
> 
> I should say sorry that your PR stayed without a review, problems with rule were not on my priority list, but the main reason is that you did not provide a short test for the problem.
> 
> I have a patch that solves the problem without big changes ([Kojoley@fa0483f](https://github.com/Kojoley/spirit/commit/fa0483fd49231f1d3fba51ff6cdf4022f0401f31)), I will prepare a PR after dealing with #453.

Just FYI, after cloning your repository and doing:
```bash
evansl@lje-DL:~/prog_dev/boost/releases/ro/boost_1_69_0/sandbox/kojoley/Kololey-spirit$ git checkout --track origin/x3-perform-transformation-before-parse_rule
Branch 'x3-perform-transformation-before-parse_rule' set up to track remote branch 'x3-perform-transformation-before-parse_rule' from 'origin'.
Switched to a new branch 'x3-perform-transformation-before-parse_rule'
evansl@lje-DL:~/prog_dev/boost/releases/ro/boost_1_69_0/sandbox/kojoley/Kololey-spirit$ 
```
Tried testing:
```bash
evansl@lje-DL:~/prog_dev/boost/releases/ro/boost_1_69_0/sandbox/kojoley/Kololey-spirit/test/x3$ b2 x3_rule_separate_tu
sh: 1: icpc: not found
Performing configuration checks

    - default address-model    : 64-bit (cached)
    - default architecture     : x86 (cached)
    - Boost.Config Feature Check: cxx14_decltype_auto : yes (cached)
    - Boost.Config Feature Check: cxx14_generic_lambdas : yes (cached)
    - Boost.Config Feature Check: cxx14_return_type_deduction : yes (cached)
    - Boost.Config Feature Check: cxx14_variable_templates : yes (cached)
    - Boost.Config Feature Check: cxx14_decltype_auto : yes (cached)
    - Boost.Config Feature Check: cxx14_generic_lambdas : yes (cached)
    - Boost.Config Feature Check: cxx14_return_type_deduction : yes (cached)
    - Boost.Config Feature Check: cxx14_variable_templates : yes (cached)
...patience...
...patience...
...found 1599 targets...
...updating 3 targets...
clang-linux.link ../../../../../bin.v2/sandbox/kojoley/Kololey-spirit/test/x3/x3_rule_separate_tu.test/clang-linux-6.0/debug/visibility-hidden/x3_rule_separate_tu
../../../../../bin.v2/sandbox/kojoley/Kololey-spirit/test/x3/x3_rule_separate_tu.test/clang-linux-6.0/debug/visibility-hidden/rule_separate_tu.o: In function `bool boost::spirit::x3::rule<unused_attr::grammar_r, boost::spirit::x3::unused_type, false>::parse<char const*, boost::spirit::x3::unused_type, boost::spirit::x3::unused_type const>(char const*&, char const* const&, boost::spirit::x3::unused_type const&, boost::spirit::x3::unused_type, boost::spirit::x3::unused_type const&) const':
/home/evansl/prog_dev/boost/releases/ro/boost_1_69_0/sandbox/kojoley/Kololey-spirit/test/x3/../../../../../boost/spirit/home/x3/nonterminal/rule.hpp:113: undefined reference to `bool unused_attr::parse_rule<char const*, boost::spirit::x3::unused_type, boost::spirit::x3::unused_type const>(boost::spirit::x3::rule<unused_attr::grammar_r, boost::spirit::x3::unused_type, false>, char const*&, char const* const&, boost::spirit::x3::unused_type const&, boost::spirit::x3::unused_type const&)'

{snip}
```

--- a/include/boost/spirit/home/x3/nonterminal/rule.hpp
+++ b/include/boost/spirit/home/x3/nonterminal/rule.hpp
@@ -114,7 +114,17 @@ namespace boost { namespace spirit { namespace x3
             static_assert(has_attribute,
                 "The rule does not have an attribute. Check your parser.");
 
-            return parse_rule(*this, first, last, context, attr);
+            using transform = traits::transform_attribute<
+                Attribute_, attribute_type, parser_id>;
+
+            using transform_attr = typename transform::type;
+            transform_attr attr_ = transform::pre(attr);
+
+            if (parse_rule(*this, first, last, context, attr_)) {
+                transform::post(attr, std::forward<transform_attr>(attr_));
+                return true;
+            }
+            return false;
         }
 
         template <typename Iterator, typename Context>

--- a/test/x3/rule3.cpp
+++ b/test/x3/rule3.cpp
@@ -38,6 +38,19 @@ struct stationary : boost::noncopyable
 };
 
 
+namespace check_stationary {
+
+boost::spirit::x3::rule<class a_r, stationary> const a;
+boost::spirit::x3::rule<class b_r, stationary> const b;
+
+auto const a_def = '{' >> boost::spirit::x3::int_ >> '}';
+auto const b_def = a;
+
+BOOST_SPIRIT_DEFINE(a, b)
+
+}
+
+
 int main()
 {
     using spirit_test::test_attr;
@@ -46,7 +59,6 @@ int main()
     using namespace boost::spirit::x3::ascii;
     using boost::spirit::x3::rule;
     using boost::spirit::x3::lit;
-    using boost::spirit::x3::int_;
     using boost::spirit::x3::eps;
     using boost::spirit::x3::unused_type;
 
@@ -91,11 +103,8 @@ int main()
     }
 
     { // ensure no unneded synthesization, copying and moving occured
-        auto a = rule<class a_r, stationary>{} = '{' >> int_ >> '}';
-        auto b = rule<class b_r, stationary>{} = a;
-
         stationary st { 0 };
-        BOOST_TEST(test_attr("{42}", b, st));
+        BOOST_TEST(test_attr("{42}", check_stationary::b, st));
         BOOST_TEST_EQ(st.val, 42);
     }
 

--- a/test/x3/rule_separate_tu.cpp
+++ b/test/x3/rule_separate_tu.cpp
@@ -26,7 +26,9 @@ int main()
     }
 
     {
-        int i;
+        long i;
+        static_assert(!std::is_same<decltype(i), used_attr::grammar_type::attribute_type>::value,
+            "ensure we have instantiated the rule with a different attribute type");
         BOOST_TEST(test_attr("123", used_attr::grammar, i));
         BOOST_TEST_EQ(i, 123);
         BOOST_TEST(test_attr(" 42", used_attr::grammar, i, used_attr::skipper));

